File: src/lib/data/models/analytics.dart
import 'json.dart';
import 'money.dart';

/// `GET /analytics/overview`.
class AnalyticsOverview {
  final int totalWorks;
  final int green;
  final int yellow;
  final int red;
  final double averageRiskScore;
  final Money totalSanctioned;
  final int pendingFundReleases;
  final int totalEvidenceItems;
  final int totalDecisions;

  const AnalyticsOverview({
    required this.totalWorks,
    required this.green,
    required this.yellow,
    required this.red,
    required this.averageRiskScore,
    required this.totalSanctioned,
    required this.pendingFundReleases,
    required this.totalEvidenceItems,
    required this.totalDecisions,
  });

  factory AnalyticsOverview.fromJson(Map<String, dynamic> json) {
    final bands = asMap(json['by_band']);
    return AnalyticsOverview(
      totalWorks: asInt(json['total_works']),
      green: asInt(bands['green']),
      yellow: asInt(bands['yellow']),
      red: asInt(bands['red']),
      averageRiskScore: asDouble(json['average_risk_score']),
      totalSanctioned: Money.fromJson(json['total_sanctioned_paise']),
      pendingFundReleases: asInt(json['pending_fund_releases']),
      totalEvidenceItems: asInt(json['total_evidence_items']),
      totalDecisions: asInt(json['total_decisions']),
    );
  }

  /// Works that actually carry an assessment. A work with no assessment is not
  /// a low-risk work, so the pie is drawn over this, not over [totalWorks].
  int get assessedWorks => green + yellow + red;

  bool get hasAssessments => assessedWorks > 0;
}

/// `GET /analytics/category-risk`.
class CategoryRisk {
  final String category;
  final int count;
  final double avgScore;
  final int maxScore;

  const CategoryRisk({
    required this.category,
    required this.count,
    required this.avgScore,
    required this.maxScore,
  });

  factory CategoryRisk.fromJson(Map<String, dynamic> json) => CategoryRisk(
        category: asString(json['category']),
        count: asInt(json['count']),
        avgScore: asDouble(json['avg_score']),
        maxScore: asInt(json['max_score']),
      );
}

/// `GET /analytics/district-summary`.
class DistrictSummary {
  final String district;
  final int works;
  final Money totalSanctioned;
  final double avgRiskScore;

  const DistrictSummary({
    required this.district,
    required this.works,
    required this.totalSanctioned,
    required this.avgRiskScore,
  });

  factory DistrictSummary.fromJson(Map<String, dynamic> json) => DistrictSummary(
        district: asString(json['district']),
        works: asInt(json['works']),
        totalSanctioned: Money.fromJson(json['total_sanctioned_paise']),
        avgRiskScore: asDouble(json['avg_risk_score']),
      );
}

/// `GET /analytics/top-risk`.
class TopRiskWork {
  final String workCode;
  final String title;
  final String category;
  final int score;
  final String band;

  const TopRiskWork({
    required this.workCode,
    required this.title,
    required this.category,
    required this.score,
    required this.band,
  });

  factory TopRiskWork.fromJson(Map<String, dynamic> json) => TopRiskWork(
        workCode: asString(json['work_code']),
        title: asString(json['title']),
        category: asString(json['category']),
        score: asInt(json['score']),
        band: asString(json['band']),
      );
}

/// `GET /analytics/rule-frequency` — which rules actually fire in this dataset.
class RuleFrequency {
  final String code;
  final String category;
  final int fires;
  final double avgPoints;

  const RuleFrequency({
    required this.code,
    required this.category,
    required this.fires,
    required this.avgPoints,
  });

  factory RuleFrequency.fromJson(Map<String, dynamic> json) => RuleFrequency(
        code: asString(json['code']),
        category: asString(json['category']),
        fires: asInt(json['fires']),
        avgPoints: asDouble(json['avg_points']),
      );
}

/// `GET /readyz` — the app shows the demo flag rather than hiding it.
class ServerStatus {
  final String status;
  final String engineVersion;
  final String rulesSha256;
  final bool demoMode;
  final String env;
  final String? db;

  const ServerStatus({
    required this.status,
    required this.engineVersion,
    required this.rulesSha256,
    required this.demoMode,
    required this.env,
    this.db,
  });

  factory ServerStatus.fromJson(Map<String, dynamic> json) => ServerStatus(
        status: asString(json['status']),
        engineVersion: asString(json['engine_version']),
        rulesSha256: asString(json['rules_sha256']),
        demoMode: asBool(json['demo_mode']),
        env: asString(json['env']),
        db: asStringOrNull(json['db']),
      );

  String get shortRulesSha =>
      rulesSha256.length >= 12 ? rulesSha256.substring(0, 12) : rulesSha256;
}

/// `GET /audit/verify` — the chain integrity check, run live on stage.
class AuditVerification {
  final bool valid;
  final int checked;
  final int? firstSeq;
  final int? lastSeq;
  final String? headHash;
  final int? brokenAtSeq;
  final String? problem;

  const AuditVerification({
    required this.valid,
    required this.checked,
    this.firstSeq,
    this.lastSeq,
    this.headHash,
    this.brokenAtSeq,
    this.problem,
  });

  factory AuditVerification.fromJson(Map<String, dynamic> json) => AuditVerification(
        valid: asBool(json['valid']),
        checked: asInt(json['checked']),
        firstSeq: asIntOrNull(json['first_seq']),
        lastSeq: asIntOrNull(json['last_seq']),
        headHash: asStringOrNull(json['head_hash']),
        brokenAtSeq: asIntOrNull(json['broken_at_seq']),
        problem: asStringOrNull(json['problem']),
      );
}

File: server/app/routers/analytics.py
"""Analytics and reporting endpoints for the dashboard."""

from __future__ import annotations

from fastapi import APIRouter
from sqlalchemy import func, select

from app.deps import CurrentUser, DbSession
from app.models import (
    Decision,
    Evidence,
    FundRelease,
    RiskAssessment,
    Work,
)

router = APIRouter(prefix="/api/v1/analytics", tags=["analytics"])


@router.get("/overview", operation_id="getAnalyticsOverview")
def overview(db: DbSession, user: CurrentUser):
    """Top-level numbers for the admin dashboard."""
    total = db.execute(
        select(func.count()).select_from(Work).where(Work.deleted_at.is_(None))
    ).scalar_one()

    by_band = dict(
        db.execute(
            select(RiskAssessment.band, func.count())
            .where(RiskAssessment.is_current.is_(True))
            .group_by(RiskAssessment.band)
        ).all()
    )

    avg_score = db.execute(
        select(func.avg(RiskAssessment.score)).where(RiskAssessment.is_current.is_(True))
    ).scalar_one()

    total_sanctioned = db.execute(
        select(func.sum(Work.sanctioned_amount_paise)).where(Work.deleted_at.is_(None))
    ).scalar_one() or 0

    pending_releases = db.execute(
        select(func.count()).select_from(FundRelease).where(FundRelease.status == "pending")
    ).scalar_one()

    total_evidence = db.execute(
        select(func.count()).select_from(Evidence).where(Evidence.deleted_at.is_(None))
    ).scalar_one()

    total_decisions = db.execute(
        select(func.count()).select_from(Decision)
    ).scalar_one()

    return {
        "total_works": total,
        "by_band": {
            "green": by_band.get("green", 0),
            "yellow": by_band.get("yellow", 0),
            "red": by_band.get("red", 0),
        },
        # None when nothing has been scored yet, never 0. A portfolio with no
        # assessments is not a portfolio averaging zero risk, and 0 is the one
        # value a reader would take as "all clear". The client already draws
        # its distribution over the assessed count rather than total_works;
        # this keeps the API telling the same truth to anyone else reading it.
        "average_risk_score": None if avg_score is None else round(float(avg_score), 1),
        "assessed_works": sum(by_band.values()),
        "total_sanctioned_paise": total_sanctioned,
        "pending_fund_releases": pending_releases,
        "total_evidence_items": total_evidence,
        "total_decisions": total_decisions,
    }


@router.get("/category-risk", operation_id="getCategoryRiskDistribution")
def category_risk(db: DbSession, user: CurrentUser):
    """Risk distribution by work category."""
    rows = db.execute(
        select(
            Work.category,
            func.count().label("count"),
            func.avg(RiskAssessment.score).label("avg_score"),
            func.max(RiskAssessment.score).label("max_score"),
        )
        .join(RiskAssessment, Work.current_assessment_id == RiskAssessment.id)
        .where(Work.deleted_at.is_(None))
        .group_by(Work.category)
        .order_by(func.avg(RiskAssessment.score).desc())
    ).all()

    return [
        {
            "category": r.category,
            "count": r.count,
            "avg_score": round(float(r.avg_score), 1) if r.avg_score else 0,
            "max_score": r.max_score or 0,
        }
        for r in rows
    ]


@router.get("/district-summary", operation_id="getDistrictSummary")
def district_summary(db: DbSession, user: CurrentUser):
    """Per-district aggregates."""
    from app.models import District
    rows = db.execute(
        select(
            District.name,
            func.count(Work.id).label("works"),
            # The average covers assessed works only, so the count of those
            # has to travel with it. Reporting "657 works, average risk 6.8"
            # when 50 were assessed states something that is not true.
            func.count(RiskAssessment.id).label("assessed"),
            func.sum(Work.sanctioned_amount_paise).label("total_paise"),
            func.avg(RiskAssessment.score).label("avg_score"),
        )
        .join(Work, Work.district_id == District.id)
        .outerjoin(RiskAssessment, Work.current_assessment_id == RiskAssessment.id)
        .where(Work.deleted_at.is_(None))
        .group_by(District.name)
    ).all()

    return [
        {
            "district": r.name,
            "works": r.works,
            "assessed_works": r.assessed,
            "total_sanctioned_paise": int(r.total_paise or 0),
            # None, not 0. This is an OUTER join: a district whose works have
            # no current assessment yields NULL here, and reporting that as 0
            # would paint an entirely unassessed district as the safest one on
            # the screen.
            "avg_risk_score": (
                round(float(r.avg_score), 1) if r.avg_score is not None else None
            ),
        }
        for r in rows
    ]


@router.get("/top-risk", operation_id="getTopRiskWorks")
def top_risk(db: DbSession, user: CurrentUser, limit: int = 20):
    """Top N highest-risk works."""
    rows = db.execute(
        select(Work.work_code, Work.title, Work.category, RiskAssessment.score, RiskAssessment.band)
        .join(RiskAssessment, Work.current_assessment_id == RiskAssessment.id)
        .where(Work.deleted_at.is_(None))
        .order_by(RiskAssessment.score.desc())
        .limit(limit)
    ).all()

    return [
        {
            "work_code": r.work_code,
            "title": r.title,
            "category": r.category,
            "score": r.score,
            "band": r.band,
        }
        for r in rows
    ]


@router.get("/rule-frequency", operation_id="getRuleFrequency")
def rule_frequency(db: DbSession, user: CurrentUser):
    """How often each rule fires across all current assessments."""
    from app.models import RiskReason
    rows = db.execute(
        select(
            RiskReason.code,
            RiskReason.category,
            func.count().label("fires"),
            func.avg(RiskReason.points).label("avg_points"),
        )
        .join(RiskAssessment, RiskReason.assessment_id == RiskAssessment.id)
        .where(RiskAssessment.is_current.is_(True), RiskReason.points > 0)
        .group_by(RiskReason.code, RiskReason.category)
        .order_by(func.count().desc())
    ).all()

    return [
        {
            "code": r.code,
            "category": r.category,
            "fires": r.fires,
            "avg_points": round(float(r.avg_points), 1),
        }
        for r in rows
    ]
